plugin-maven/src/main/java/com/diffplug/spotless/maven/incremental/PluginFingerprint.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2021 DiffPlug
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.diffplug.spotless.maven.incremental;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.io.UncheckedIOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Base64;
+import java.util.Objects;
+
+import org.apache.maven.model.Plugin;
+import org.apache.maven.project.MavenProject;
+
+class PluginFingerprint {
+
+	private static final String SPOTLESS_PLUGIN_KEY = "com.diffplug.spotless:spotless-maven-plugin";
+
+	private final String value;
+
+	private PluginFingerprint(String value) {
+		this.value = value;
+	}
+
+	static PluginFingerprint from(MavenProject project) {
+		Plugin spotlessPlugin = project.getPlugin(SPOTLESS_PLUGIN_KEY);
```

Comment:
@nedtwigg I think https://github.com/diffplug/spotless/pull/935/commits/9bc59fe756795988e65b71f841e2fc51be0f3514 should address this problem. It makes `PluginFingerprint` hash the `Plugin` object and all the `Formatter`s. Class `FormattersHolder` simplifies creation/closing of multiple formatters. `ObjectDigestOutputStream` simplifies serialization of multiple objects into a single digest. The code is still not production-ready and tests are missing. Could you please take a look if the general approach makes sense?

Reply:
I think you can leave the `Plugin` object out and just do the `Formatters`. IMO you should only keep the `Plugin` if you can point to a specific thing that isn't being captured without it. I just surfed the manual and couldn't find anything, except...

One thing missing is the `GitRatchet` state. In the Gradle plugin we handle it like so:

https://github.com/diffplug/spotless/blob/f9d66c4dc71562a36f107ccc3acd3455c3de835d/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java#L105-L108

https://github.com/diffplug/spotless/blob/f9d66c4dc71562a36f107ccc3acd3455c3de835d/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTask.java#L89-L93

This state matters because it affects whether the `Formatter` runs at all or not.

https://github.com/diffplug/spotless/blob/f9d66c4dc71562a36f107ccc3acd3455c3de835d/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessTaskImpl.java#L68-L72

If you'd like, you can do a check that fails with an error like "Can't use git ratchet at the same time as incremental up-to-date, pick just one", and I can submit a PR to you later to fix it.

Reply:
Second thought: I just made a comment about needing `ObjectId subtreeSha` and another comment about replacing `List<File>` with `Supplier<Iterable<File>>`. I think I'm probably wrong about both (or maybe neither, let me explain 😆 )

I linked to the Gradle plugin sourcecode above. Looking at the maven plugin, the GitRatchet works like so:

https://github.com/diffplug/spotless/blob/2d3acdea75d08b82f3181f975e2de099552c28c6/plugin-maven/src/main/java/com/diffplug/spotless/maven/AbstractSpotlessMojo.java#L186-L194

So I think that maybe what you have right now is okay - do up-to-dating based on an eager `List<File>`, keeping in mind that that list will change based on git state which is handled by `GitRatchet`. If you took my feedback of adding `subtreeSha` into the state, then you could maybe also replace `List<File>` with a lazy `Supplier<Iterable<File>>`. But since you are responsible for doing the up-to-date check on every file yourself (gradle gives that to you for "free"), then maybe that doesn't make sense.

Carry-on, ignore me :)

Reply:
TL;DR: my answer to

> Could you please take a look if the general approach makes sense?

is yes! It looks good and I think it will work great.

Reply:
Thanks for the great input! :)

 1. I kept `Plugin` in the fingerprint SHA to make the fingerprint detect plugin version changes and maybe plugin dependency changes.
    Plugin version change example:
    ```diff
    <plugin>
      <groupId>com.diffplug.spotless</groupId>
      <artifactId>spotless-maven-plugin</artifactId>
    -  <version>2.17.3</version>
    +  <version>3.0.0</version>
      <configuration>
      ...
      </configuration>
    </plugin>
    ```
    Contrived plugin dependency change example:
    ```diff
    <plugin>
      <groupId>com.diffplug.spotless</groupId>
      <artifactId>spotless-maven-plugin</artifactId>
      <version>2.17.3</version>
      <configuration>
      ...
      </configuration>
    +  <dependencies>
    +   <!-- Override durian-core 1.2.0 used by the plugin to workaround a bug -->
    +   <dependency>
    +     <groupId>com.diffplug.durian</groupId>
    +     <artifactId>durian-core</artifactId>
    +     <version>1.3.0</version>
    +    </dependency>
    +  </dependencies>
    </plugin>
    ```

    Does it make sense to capture plugin version and dependencies in `PluginFingerptint`? Would serialized formatters capture plugin version and dependencies?

 2. Right now up-to-date checking is applied to files returned by the `#collectFiles()` method:
https://github.com/diffplug/spotless/blob/2d3acdea75d08b82f3181f975e2de099552c28c6/plugin-maven/src/main/java/com/diffplug/spotless/maven/AbstractSpotlessMojo.java#L186-L194
This method gets files from either `GitRatchet` or from include/exclude filename masks. Up-to-date checking is applied after ratchet-ing. Thus maybe `PluginFingerprint` does not need to include JGit `ObjectId`?

 3. I think it should be possible to change `List<File>` to `Supplier<Iterable<File>>`. The list of files is used to create a `LineEnding.Policy` for each formatter: https://github.com/diffplug/spotless/blob/f86bb04da7704b89b74fc6bf25d4218fcfe604be/plugin-maven/src/main/java/com/diffplug/spotless/maven/FormatterFactory.java#L74-L77 It could be beneficial to delay the creation of `List<File>` until either `LineEnding.Policy` needs it or `Formatter` needs to format. This way the plugin will only keep a single `List<File>` in memory that's needed by the current formatter. I can change the code to address this, if you agree.

Reply:
> 1. I kept Plugin in the fingerprint SHA to make the fingerprint detect plugin version changes 

Good call, you should keep it how you have it.

> 2. git ratchet

I think we should make a test and see. I can help with that when you're ready, just lmk.

> 3. I can change the code to address this, if you agree.

I agree :)